File: hdmi-capture/led_layout.py
"""
Android tarafındaki `LedLayoutConfig` (Kotlin) ile birebir aynı modelin
Python portu. Her kenarın kendi LED sayısı ve açık/kapalı durumu vardır;
bu sayede tam perimetre, yalnızca üst-alt veya L-şekilli gibi kurulumlar
desteklenir.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class Edge(Enum):
    TOP = "TOP"
    RIGHT = "RIGHT"
    BOTTOM = "BOTTOM"
    LEFT = "LEFT"


@dataclass
class EdgeConfig:
    enabled: bool = True
    led_count: int = 0


@dataclass
class LedLayoutConfig:
    start_edge: Edge = Edge.TOP
    top: EdgeConfig = field(default_factory=lambda: EdgeConfig(True, 20))
    right: EdgeConfig = field(default_factory=lambda: EdgeConfig(True, 10))
    bottom: EdgeConfig = field(default_factory=lambda: EdgeConfig(True, 20))
    left: EdgeConfig = field(default_factory=lambda: EdgeConfig(True, 10))

    def config_for(self, edge: Edge) -> EdgeConfig:
        return {
            Edge.TOP: self.top,
            Edge.RIGHT: self.right,
            Edge.BOTTOM: self.bottom,
            Edge.LEFT: self.left,
        }[edge]

    def wiring_order(self) -> list[Edge]:
        canonical = [Edge.TOP, Edge.RIGHT, Edge.BOTTOM, Edge.LEFT]
        start_index = canonical.index(self.start_edge)
        rotated = canonical[start_index:] + canonical[:start_index]
        return [e for e in rotated if self.config_for(e).enabled and self.config_for(e).led_count > 0]

    def total_led_count(self) -> int:
        return sum(self.config_for(e).led_count for e in self.wiring_order())

    @staticmethod
    def from_dict(data: dict | None) -> "LedLayoutConfig":
        data = data or {}

        def edge_cfg(key: str, default_enabled: bool, default_count: int) -> EdgeConfig:
            d = data.get(key, {})
            return EdgeConfig(d.get("enabled", default_enabled), d.get("led_count", default_count))

        return LedLayoutConfig(
            start_edge=Edge(data.get("start_edge", "TOP")),
            top=edge_cfg("top", True, 20),
            right=edge_cfg("right", True, 10),
            bottom=edge_cfg("bottom", True, 20),
            left=edge_cfg("left", True, 10),
        )

File: app/src/main/java/com/ambiscreen/app/capture/LedLayout.kt
package com.ambiscreen.app.capture

enum class Edge { TOP, RIGHT, BOTTOM, LEFT }

data class EdgeConfig(
    val enabled: Boolean = true,
    val ledCount: Int = 0,
)

/**
 * Fiziksel LED şeridinin ekran çevresindeki yerleşimini tanımlar. Her kenarın
 * kendi LED sayısı ve açık/kapalı durumu vardır; bu sayede tam perimetre
 * (4 kenar), yalnızca üst-alt, L-şekilli (örn. üst+sol) veya tek kenar gibi
 * kurulumlar desteklenir. `startEdge`, LED şeridinin fiziksel olarak hangi
 * kenardan başladığını belirtir; açık kenarlar bu noktadan saat yönünde
 * sırayla dizilip DDP'ye o sırayla gönderilir.
 */
data class LedLayoutConfig(
    val startEdge: Edge = Edge.TOP,
    val top: EdgeConfig = EdgeConfig(true, 20),
    val right: EdgeConfig = EdgeConfig(true, 10),
    val bottom: EdgeConfig = EdgeConfig(true, 20),
    val left: EdgeConfig = EdgeConfig(true, 10),
) {
    fun configFor(edge: Edge): EdgeConfig = when (edge) {
        Edge.TOP -> top
        Edge.RIGHT -> right
        Edge.BOTTOM -> bottom
        Edge.LEFT -> left
    }

    fun with(edge: Edge, config: EdgeConfig): LedLayoutConfig = when (edge) {
        Edge.TOP -> copy(top = config)
        Edge.RIGHT -> copy(right = config)
        Edge.BOTTOM -> copy(bottom = config)
        Edge.LEFT -> copy(left = config)
    }

    /** Saat yönünde, startEdge'den başlayarak yalnızca açık ve LED'i olan kenarları sıralar. */
    fun wiringOrder(): List<Edge> {
        val canonical = listOf(Edge.TOP, Edge.RIGHT, Edge.BOTTOM, Edge.LEFT)
        val startIndex = canonical.indexOf(startEdge).coerceAtLeast(0)
        val rotated = canonical.subList(startIndex, canonical.size) + canonical.subList(0, startIndex)
        return rotated.filter { configFor(it).enabled && configFor(it).ledCount > 0 }
    }

    fun totalLedCount(): Int = wiringOrder().sumOf { configFor(it).ledCount }

    fun serialize(): String {
        fun encode(cfg: EdgeConfig) = "${if (cfg.enabled) 1 else 0}:${cfg.ledCount}"
        return "${startEdge.name}|${encode(top)}|${encode(right)}|${encode(bottom)}|${encode(left)}"
    }

    companion object {
        val DEFAULT = LedLayoutConfig()

        fun deserialize(raw: String?): LedLayoutConfig {
            if (raw.isNullOrBlank()) return DEFAULT
            return try {
                val parts = raw.split("|")
                fun parse(part: String): EdgeConfig {
                    val (enabledFlag, count) = part.split(":")
                    return EdgeConfig(enabledFlag == "1", count.toInt())
                }
                LedLayoutConfig(
                    startEdge = Edge.valueOf(parts[0]),
                    top = parse(parts[1]),
                    right = parse(parts[2]),
                    bottom = parse(parts[3]),
                    left = parse(parts[4]),
                )
            } catch (e: Exception) {
                DEFAULT
            }
        }
    }
}
